core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -135,52 +141,53 @@ public void start() {
          if (transactionManager != null) {
             xaTx = transactionManager.suspend();
          }
-         storesMutex.readLock().lock();
+         storesMutex.writeLock().lock();
          try {
+            Set<Lifecycle> undelegated = new HashSet<>();
+            nonTxWriters.forEach(w -> startWriter(w, undelegated));
+            txWriters.forEach(w -> startWriter(w, undelegated));
+            loaders.forEach(l -> startLoader(l, undelegated));
 
-            Set<Lifecycle> undelegated = new HashSet<>();//black magic to make sure the store start only gets invoked once
+            // Ensure that after writers and loaders have started, they are classified as available by their isAvailable impl
+            pollStoreAvailability();
 
-            Consumer<CacheWriter> startWriter = writer -> {
-               writer.start();
-               if (writer instanceof DelegatingCacheWriter) {
-                  CacheWriter actual = undelegate(writer);
-                  actual.start();
-                  undelegated.add(actual);
-               } else {
-                  undelegated.add(writer);
-               }
-
-               if (configMap.get(writer).purgeOnStartup()) {
-                  if (!(writer instanceof AdvancedCacheWriter))
-                     throw new PersistenceException("'purgeOnStartup' can only be set on stores implementing " +
-                                                          "" + AdvancedCacheWriter.class.getName());
-                  ((AdvancedCacheWriter) writer).clear();
-               }
-            };
-            nonTxWriters.forEach(startWriter);
-            txWriters.forEach(startWriter);
-
-            for (CacheLoader l : loaders) {
-               if (!undelegated.contains(l))
-                  l.start();
-               if (l instanceof DelegatingCacheLoader) {
-                  CacheLoader actual = undelegate(l);
-                  if (!undelegated.contains(actual)) {
-                     actual.start();
-                  }
-               }
-            }
+            // Now schedule the availability check
+            long interval = configuration.persistence().availabilityInterval();
+            persistenceExecutor.scheduleAtFixedRate(this::pollStoreAvailability, interval, interval, TimeUnit.MILLISECONDS);
```

Comment:
We should retain a reference to the `Future` so we can stop it when the `stop` method is invoked or when all cache stores have been disabled since there is no way to add a new store to a running system.